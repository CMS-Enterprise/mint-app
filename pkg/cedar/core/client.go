package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
