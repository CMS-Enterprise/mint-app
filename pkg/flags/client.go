package flags

import (
	"context"
	"crypto/sha256"
	"fmt"
	"time"

	"github.com/cmsgov/mint-app/pkg/appcontext"

	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfiledata"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfilewatch"

	"github.com/cmsgov/mint-app/pkg/appconfig"
)

// DowngradeAssessmentTeamKey is the Flag Key in LaunchDarkly for downgrading assessment team users
const DowngradeAssessmentTeamKey = "downgradeAssessmentTeam"

// Config has all the parts for creating a new LD Client
type Config struct {
	Source         appconfig.FlagSourceOption
	Key            string
	Timeout        time.Duration
	FlagValuesFile string
}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	if config.Source == appconfig.FlagSourceLaunchDarkly {
		return ld.MakeClient(config.Key, config.Timeout)
	}

	defaultTimeout := 5 * time.Second

	if config.Source == appconfig.FlagSourceFile {

		ldConfig := ld.Config{
			DataSource: ldfiledata.DataSource().FilePaths(config.FlagValuesFile).Reloader(ldfilewatch.WatchFiles),
		}
		return ld.MakeCustomClient("fake_key", ldConfig, defaultTimeout)
	}

	// we default to an OFFLINE client for non-deployed environments
	return ld.MakeCustomClient("fake_offline_key", ld.Config{Offline: true}, 5*time.Second)
}

// Principal builds the LaunchDarkly user object for the
// currently authenticated principal.
func Principal(ctx context.Context) lduser.User {
	p := appcontext.Principal(ctx)
	key := UserKeyForID(p.ID())

	// this is a bit of a loose inference, assuming a user w/o Job Codes
	// is an Anonymous user. Over time, may want to consider adding
	// a `func Anonymous() bool` accessor to the authorizaion.Principal interface
	// definition instead of doing this inference
	authed := (p.AllowUSER() || p.AllowASSESSMENT())

	return lduser.
		NewUserBuilder(key).
		Anonymous(!authed).
		Build()
}

// UserKeyForID generates a user key from an ID
// we should not be using bare EUA IDs as identifiers to
// LaunchDarkly (per Jimil/ISSO), so we use a cryptographically
// secure one-way hash of the EUA ID as "key" for the LD User object.
func UserKeyForID(id string) string {
	h := sha256.New()
	_, _ = h.Write([]byte(id))
	return fmt.Sprintf("%x", h.Sum(nil))
}
