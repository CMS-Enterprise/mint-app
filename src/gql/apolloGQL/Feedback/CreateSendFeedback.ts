import { graphql } from '../../gen/gql';

export default graphql(/* GraphQL */ `
  mutation CreatSendFeedback($input: SendFeedbackEmailInput!) {
    sendFeedbackEmail(input: $input)
  }
`);
