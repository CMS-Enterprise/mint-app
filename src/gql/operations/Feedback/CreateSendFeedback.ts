import { gql } from '@apollo/client';

export default gql(/* GraphQL */ `
  mutation CreatSendFeedback($input: SendFeedbackEmailInput!) {
    sendFeedbackEmail(input: $input)
  }
`);
