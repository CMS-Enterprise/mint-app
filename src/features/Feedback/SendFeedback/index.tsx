import React, { Fragment } from 'react';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';
import { useMutation } from '@apollo/client';
import {
  Button,
  CharacterCount,
  Fieldset,
  Grid,
  GridContainer,
  Label,
  Radio,
  TextInput
} from '@trussworks/react-uswds';
import HelpBreadcrumb from 'features/HelpAndKnowledge/Articles/_components/HelpBreadcrumb';
import { Field, Formik, FormikProps } from 'formik';
import {
  EaseOfUse,
  MintUses,
  SatisfactionLevel,
  SendFeedbackEmailInput
} from 'gql/generated/graphql';
import CreateSendFeedback from 'gql/operations/Feedback/CreateSendFeedback';

import BooleanRadio from 'components/BooleanRadioForm';
import CheckboxField from 'components/CheckboxField';
import FieldGroup from 'components/FieldGroup';
import MainContent from 'components/MainContent';
import MINTForm from 'components/MINTForm';
import PageHeading from 'components/PageHeading';
import { useErrorMessage } from 'contexts/ErrorContext';
import { getKeys } from 'types/translation';
import { tObject } from 'utils/translation';

const SendFeedback = () => {
  const { t } = useTranslation(['feedback', 'miscellaneous']);

  const navigate = useNavigate();

  const { setErrorMeta } = useErrorMessage();

  const [update, { loading }] = useMutation(CreateSendFeedback);

  const mintUsedForOptions = tObject<MintUses>('feedback:mintUsedFor.options');

  const systemEasyToUseOptions = tObject<EaseOfUse>(
    'feedback:systemEasyToUse.options'
  );

  const howSatisfiedOptions = tObject<SatisfactionLevel>(
    'feedback:howSatisfied.options'
  );

  const handleFormSubmit = (formikValues: SendFeedbackEmailInput) => {
    setErrorMeta({
      overrideMessage: t('errorFeedback')
    });

    update({
      variables: {
        input: formikValues
      }
    })
      .then(response => {
        if (!response?.errors) {
          navigate('/feedback-received');
        }
      })
      .catch(errors => {
        window.scrollTo(0, 0);
      });
  };

  const initialValues: SendFeedbackEmailInput = {
    isAnonymousSubmission: false,
    allowContact: null,
    cmsRole: '',
    mintUsedFor: [],
    mintUsedForOther: '',
    systemEasyToUse: null,
    systemEasyToUseOther: '',
    howSatisfied: null
  };

  return (
    <MainContent>
      <GridContainer>
        <HelpBreadcrumb newTabOnly />

        <PageHeading className="margin-bottom-2">
          {t('feedbackHeading')}
        </PageHeading>

        <p className="margin-bottom-2 font-body-lg  line-height-sans-lg">
          {t('feedbackSubheading')}
        </p>

        <Formik
          initialValues={initialValues}
          onSubmit={values => {
            handleFormSubmit(values);
          }}
          enableReinitialize
        >
          {(formikProps: FormikProps<SendFeedbackEmailInput>) => {
            const { handleSubmit, setFieldValue, values } = formikProps;

            return (
              <Grid desktop={{ col: 6 }}>
                <MINTForm
                  className="margin-top-3"
                  onSubmit={(e: React.FormEvent<HTMLFormElement>) => {
                    handleSubmit(e);
                  }}
                >
                  <Fieldset disabled={loading}>
                    <FieldGroup>
                      <Label htmlFor="send-feedback-allow-anon-submission">
                        {t('isAnonymousSubmission.label')}
                      </Label>

                      <p className="text-base margin-y-1">
                        {t('isAnonymousSubmission.sublabel')}
                      </p>

                      <BooleanRadio
                        field="isAnonymousSubmission"
                        id="send-feedback-allow-anon-submission"
                        value={values.isAnonymousSubmission}
                        setFieldValue={setFieldValue}
                        options={{
                          true: t('isAnonymousSubmission.options.true'),
                          false: t('isAnonymousSubmission.options.false')
                        }}
                      />
                    </FieldGroup>

                    <FieldGroup>
                      <Label
                        htmlFor="send-feedback-allow-contact"
                        className="margin-top-5"
                      >
                        {t('allowContact.label')}
                      </Label>

                      <p className="text-base margin-y-1">
                        {t('allowContact.sublabel')}
                      </p>

                      <BooleanRadio
                        field="allowContact"
                        id="send-feedback-allow-contact"
                        value={
                          values.allowContact === undefined
                            ? null
                            : values.allowContact
                        }
                        setFieldValue={setFieldValue}
                        disabled={values.isAnonymousSubmission === true}
                        options={{
                          true: t('allowContact.options.true'),
                          false: t('allowContact.options.false')
                        }}
                      />
                    </FieldGroup>

                    <FieldGroup>
                      <Label
                        htmlFor="send-feedback-cms-role"
                        className="margin-top-5"
                      >
                        {t('cmsRole.label')}
                      </Label>

                      <Field
                        as={TextInput}
                        id="send-feedback-cmsRole"
                        data-testid="send-feedback-cms-role"
                        name="cmsRole"
                      />
                    </FieldGroup>

                    <FieldGroup>
                      <Label
                        htmlFor="send-feedback-mint-used-for"
                        className="margin-top-5"
                      >
                        {t('mintUsedFor.label')}
                      </Label>

                      <Fieldset>
                        {getKeys(mintUsedForOptions).map(key => (
                          <Fragment key={key}>
                            <Field
                              as={CheckboxField}
                              id={`send-feedback-mint-used-for-${key}`}
                              data-testid={`send-feedback-mint-used-for-${key}`}
                              name="mintUsedFor"
                              label={t(`mintUsedFor.options.${key}`)}
                              value={key}
                              checked={values.mintUsedFor?.includes(
                                key as MintUses
                              )}
                            />

                            {key === MintUses.OTHER && (
                              <div className="margin-left-4 margin-top-1">
                                <Field
                                  as={TextInput}
                                  id="send-feedback-mint-used-for-other"
                                  data-testid="send-feedback-mint-used-for-other"
                                  disabled={
                                    !values.mintUsedFor?.includes(
                                      MintUses.OTHER
                                    )
                                  }
                                  name="mintUsedForOther"
                                />
                              </div>
                            )}
                          </Fragment>
                        ))}
                      </Fieldset>
                    </FieldGroup>

                    <FieldGroup>
                      <Label
                        htmlFor="send-feedback-ease-of-use"
                        className="margin-top-5"
                      >
                        {t('systemEasyToUse.label')}
                      </Label>

                      <Fieldset>
                        {getKeys(systemEasyToUseOptions).map(key => (
                          <Fragment key={key}>
                            <Field
                              as={Radio}
                              id={`send-feedback-ease-of-use-${key}`}
                              data-testid={`send-feedback-ease-of-use-${key}`}
                              name="systemEasyToUse"
                              label={t(`systemEasyToUse.options.${key}`)}
                              value={key}
                              checked={values.systemEasyToUse === key}
                              onChange={() => {
                                setFieldValue('systemEasyToUse', key);
                              }}
                            />
                            {key === EaseOfUse.UNSURE && (
                              <div className="margin-left-4 margin-top-1">
                                <Field
                                  as={TextInput}
                                  id="send-feedback-ease-of-use-other"
                                  data-testid="send-feedback-ease-of-use-other"
                                  disabled={
                                    values.systemEasyToUse !== EaseOfUse.UNSURE
                                  }
                                  name="systemEasyToUseOther"
                                />
                              </div>
                            )}
                          </Fragment>
                        ))}
                      </Fieldset>
                    </FieldGroup>

                    <FieldGroup>
                      <Label
                        htmlFor="send-feedback-how-satisfied"
                        className="margin-top-5"
                      >
                        {t('howSatisfied.label')}
                      </Label>

                      <Fieldset>
                        {getKeys(howSatisfiedOptions).map(key => (
                          <Fragment key={key}>
                            <Field
                              as={Radio}
                              id={`send-feedback-how-satisfied-${key}`}
                              data-testid={`send-feedback-how-satisfied-${key}`}
                              name="howSatisfied"
                              label={t(`howSatisfied.options.${key}`)}
                              value={key}
                              checked={values.howSatisfied === key}
                              onChange={() => {
                                setFieldValue('howSatisfied', key);
                              }}
                            />
                          </Fragment>
                        ))}
                      </Fieldset>
                    </FieldGroup>

                    <FieldGroup>
                      <Label
                        htmlFor="send-feedback-how-can-we-improve"
                        className="margin-top-5"
                      >
                        {t('howCanWeImprove.label')}
                      </Label>

                      <Field
                        as={CharacterCount}
                        id="send-feedback-how-can-we-improve"
                        data-testid="send-feedback-how-can-we-improve"
                        className="height-card margin-bottom-1"
                        isTextArea
                        name="howCanWeImprove"
                        maxLength={2000}
                        getCharacterCount={(text: string): number =>
                          Array.from(text).length
                        }
                      />
                    </FieldGroup>

                    <div className="margin-top-6 margin-bottom-3">
                      <Button
                        type="button"
                        className="usa-button margin-bottom-1"
                        onClick={() => {
                          handleFormSubmit(values);
                        }}
                      >
                        {t('sendFeedback')}
                      </Button>
                    </div>
                  </Fieldset>
                </MINTForm>
              </Grid>
            );
          }}
        </Formik>
      </GridContainer>
    </MainContent>
  );
};

export default SendFeedback;
