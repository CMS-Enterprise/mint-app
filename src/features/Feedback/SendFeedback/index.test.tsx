import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { act, render } from '@testing-library/react';
import { EaseOfUse, MintUses, SatisfactionLevel } from 'gql/gen/graphql';
import CreateSendFeedback from 'gql/operations/Feedback/CreateSendFeedback';

import setup from 'utils/testing/setup';

import SendFeedback from '.';

const mocks = [
  {
    request: {
      query: CreateSendFeedback,
      variables: {
        input: {
          isAnonymousSubmission: false,
          allowContact: true,
          cmsRole: 'Architect',
          mintUsedFor: [MintUses.EDIT_MODEL],
          mintUsedForOther: '',
          systemEasyToUse: EaseOfUse.UNSURE,
          systemEasyToUseOther: 'Effortless',
          howSatisfied: SatisfactionLevel.NEUTRAL,
          howCanWeImprove: 'Alot'
        }
      }
    },
    result: {
      data: {
        reportAProblem: true,
        sendFeedbackEmail: true
      }
    }
  }
];

describe('Send feedback form', () => {
  it('submits the "Send feedback" form successfully', async () => {
    await act(async () => {
      const { findByText, getByRole, getByTestId, getByText, user } = setup(
        <MemoryRouter initialEntries={['send-feedback']}>
          <Route path="send-feedback">
            <MockedProvider mocks={mocks} addTypename={false}>
              <SendFeedback />
            </MockedProvider>
          </Route>
        </MemoryRouter>
      );

      // Fill out form

      await user.click(
        getByTestId('send-feedback-allow-anon-submission-false')
      );

      await user.click(getByTestId('send-feedback-allow-contact-true'));

      await user.type(getByTestId('send-feedback-cms-role'), 'Architect');

      await user.click(getByText('To edit Model Plans'));

      await user.click(getByTestId('send-feedback-ease-of-use-UNSURE'));

      await user.type(
        getByTestId('send-feedback-ease-of-use-other'),
        'Effortless'
      );

      await user.click(getByTestId('send-feedback-how-satisfied-NEUTRAL'));

      await user.type(getByTestId('send-feedback-how-can-we-improve'), 'Alot');

      const submitButton = getByRole('button', { name: 'Send feedback' });

      await user.click(submitButton);

      // Submit success
      findByText('Thank you for your feedback');
    });
  });

  it('matches snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={['send-feedback']}>
        <Route path="send-feedback">
          <MockedProvider mocks={mocks} addTypename={false}>
            <SendFeedback />
          </MockedProvider>
        </Route>
      </MemoryRouter>
    );

    // Snapshot form submission complete state
    expect(asFragment()).toMatchSnapshot();
  });
});
