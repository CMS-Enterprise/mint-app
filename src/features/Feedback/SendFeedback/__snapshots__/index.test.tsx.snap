// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Send feedback form > matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="mint-main-content"
    id="main-content"
    role="main"
    tabindex="-1"
  >
    <div
      class="grid-container"
      data-testid="gridContainer"
    >
      <button
        class="usa-button usa-button--unstyled margin-top-6"
        data-testid="button"
        type="button"
      >
        <svg
          class="usa-icon margin-right-05 text-tbottom"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        Close tab
      </button>
      <h1
        aria-live="polite"
        class="mint-h1 margin-bottom-2"
        tabindex="-1"
      >
        Send feedback
      </h1>
      <p
        class="margin-bottom-2 font-body-lg  line-height-sans-lg"
      >
        Have a suggestion for how to improve MINT? Let us know using the form below.
      </p>
      <div
        class="desktop:grid-col-6"
        data-testid="grid"
      >
        <form
          action="#"
          class="margin-top-3"
        >
          <fieldset
            class="usa-fieldset"
            data-testid="fieldset"
          >
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label"
                data-testid="label"
                for="send-feedback-allow-anon-submission"
              >
                Would you like your feedback to remain anonymous?
              </label>
              <p
                class="text-base margin-y-1"
              >
                If you select yes, your name and email will not be recorded with your feedback.
              </p>
              <fieldset
                class="usa-fieldset"
                data-testid="fieldset"
              >
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-allow-anon-submission-true"
                    id="send-feedback-allow-anon-submission-true"
                    name="isAnonymousSubmission"
                    type="radio"
                    value="TRUE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-allow-anon-submission-true"
                  >
                    Yes
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    checked=""
                    class="usa-radio__input"
                    data-testid="send-feedback-allow-anon-submission-false"
                    id="send-feedback-allow-anon-submission-false"
                    name="isAnonymousSubmission"
                    type="radio"
                    value="FALSE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-allow-anon-submission-false"
                  >
                    No
                  </label>
                </div>
              </fieldset>
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label margin-top-5"
                data-testid="label"
                for="send-feedback-allow-contact"
              >
                May the MINT team contact you for additional information?
              </label>
              <p
                class="text-base margin-y-1"
              >
                If you choose to remain anonymous, we will be unable to contact you.
              </p>
              <fieldset
                class="usa-fieldset"
                data-testid="fieldset"
              >
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-allow-contact-true"
                    id="send-feedback-allow-contact-true"
                    name="allowContact"
                    type="radio"
                    value="TRUE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-allow-contact-true"
                  >
                    Yes
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-allow-contact-false"
                    id="send-feedback-allow-contact-false"
                    name="allowContact"
                    type="radio"
                    value="FALSE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-allow-contact-false"
                  >
                    No
                  </label>
                </div>
              </fieldset>
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label margin-top-5"
                data-testid="label"
                for="send-feedback-cms-role"
              >
                What is your role?
              </label>
              <input
                class="usa-input"
                data-testid="send-feedback-cms-role"
                id="send-feedback-cmsRole"
                name="cmsRole"
                value=""
              />
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label margin-top-5"
                data-testid="label"
                for="send-feedback-mint-used-for"
              >
                What have you used MINT for?
              </label>
              <fieldset
                class="usa-fieldset"
                data-testid="fieldset"
              >
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-VIEW_MODEL"
                    name="mintUsedFor"
                    type="checkbox"
                    value="VIEW_MODEL"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-VIEW_MODEL"
                  >
                    To view Model Plans
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-EDIT_MODEL"
                    name="mintUsedFor"
                    type="checkbox"
                    value="EDIT_MODEL"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-EDIT_MODEL"
                  >
                    To edit Model Plans
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-SHARE_MODEL"
                    name="mintUsedFor"
                    type="checkbox"
                    value="SHARE_MODEL"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-SHARE_MODEL"
                  >
                    To share or export Model Plan content
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-TRACK_SOLUTIONS"
                    name="mintUsedFor"
                    type="checkbox"
                    value="TRACK_SOLUTIONS"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-TRACK_SOLUTIONS"
                  >
                    To track operational solutions
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-CONTRIBUTE_DISCUSSIONS"
                    name="mintUsedFor"
                    type="checkbox"
                    value="CONTRIBUTE_DISCUSSIONS"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-CONTRIBUTE_DISCUSSIONS"
                  >
                    To contribute to Model Discussions
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-VIEW_HELP"
                    name="mintUsedFor"
                    type="checkbox"
                    value="VIEW_HELP"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-VIEW_HELP"
                  >
                    To view the Help Center
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="send-feedback-mint-used-for-OTHER"
                    name="mintUsedFor"
                    type="checkbox"
                    value="OTHER"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="send-feedback-mint-used-for-OTHER"
                  >
                    Other (please specify)
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <div
                  class="margin-left-4 margin-top-1"
                >
                  <input
                    class="usa-input"
                    data-testid="send-feedback-mint-used-for-other"
                    disabled=""
                    id="send-feedback-mint-used-for-other"
                    name="mintUsedForOther"
                    value=""
                  />
                </div>
              </fieldset>
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label margin-top-5"
                data-testid="label"
                for="send-feedback-ease-of-use"
              >
                The system is easy to use.
              </label>
              <fieldset
                class="usa-fieldset"
                data-testid="fieldset"
              >
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-ease-of-use-AGREE"
                    id="send-feedback-ease-of-use-AGREE"
                    name="systemEasyToUse"
                    type="radio"
                    value="AGREE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-ease-of-use-AGREE"
                  >
                    Agree
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-ease-of-use-DISAGREE"
                    id="send-feedback-ease-of-use-DISAGREE"
                    name="systemEasyToUse"
                    type="radio"
                    value="DISAGREE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-ease-of-use-DISAGREE"
                  >
                    Disagree
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-ease-of-use-UNSURE"
                    id="send-feedback-ease-of-use-UNSURE"
                    name="systemEasyToUse"
                    type="radio"
                    value="UNSURE"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-ease-of-use-UNSURE"
                  >
                    I’m not sure (please explain)
                  </label>
                </div>
                <div
                  class="margin-left-4 margin-top-1"
                >
                  <input
                    class="usa-input"
                    data-testid="send-feedback-ease-of-use-other"
                    disabled=""
                    id="send-feedback-ease-of-use-other"
                    name="systemEasyToUseOther"
                    value=""
                  />
                </div>
              </fieldset>
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label margin-top-5"
                data-testid="label"
                for="send-feedback-how-satisfied"
              >
                Overall, how satisfied are you with MINT?
              </label>
              <fieldset
                class="usa-fieldset"
                data-testid="fieldset"
              >
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-how-satisfied-VERY_SATISFIED"
                    id="send-feedback-how-satisfied-VERY_SATISFIED"
                    name="howSatisfied"
                    type="radio"
                    value="VERY_SATISFIED"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-how-satisfied-VERY_SATISFIED"
                  >
                    Very satisfied
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-how-satisfied-SATISFIED"
                    id="send-feedback-how-satisfied-SATISFIED"
                    name="howSatisfied"
                    type="radio"
                    value="SATISFIED"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-how-satisfied-SATISFIED"
                  >
                    Satisfied
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-how-satisfied-NEUTRAL"
                    id="send-feedback-how-satisfied-NEUTRAL"
                    name="howSatisfied"
                    type="radio"
                    value="NEUTRAL"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-how-satisfied-NEUTRAL"
                  >
                    Neutral
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-how-satisfied-DISSATISFIED"
                    id="send-feedback-how-satisfied-DISSATISFIED"
                    name="howSatisfied"
                    type="radio"
                    value="DISSATISFIED"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-how-satisfied-DISSATISFIED"
                  >
                    Dissatisfied
                  </label>
                </div>
                <div
                  class="usa-radio"
                  data-testid="radio"
                >
                  <input
                    class="usa-radio__input"
                    data-testid="send-feedback-how-satisfied-VERY_DISSATISFIED"
                    id="send-feedback-how-satisfied-VERY_DISSATISFIED"
                    name="howSatisfied"
                    type="radio"
                    value="VERY_DISSATISFIED"
                  />
                  <label
                    class="usa-radio__label"
                    for="send-feedback-how-satisfied-VERY_DISSATISFIED"
                  >
                    Very dissatisfied
                  </label>
                </div>
              </fieldset>
            </div>
            <div
              class="usa-form-group"
            >
              <label
                class="usa-label margin-top-5"
                data-testid="label"
                for="send-feedback-how-can-we-improve"
              >
                How can we improve MINT?
              </label>
              <textarea
                class="usa-textarea usa-character-count__field height-card margin-bottom-1"
                data-testid="send-feedback-how-can-we-improve"
                id="send-feedback-how-can-we-improve"
                name="howCanWeImprove"
              />
              <span
                class="usa-sr-only"
                id="send-feedback-how-can-we-improve-info"
              >
                You can enter up to 2000 characters
              </span>
              <div
                aria-hidden="true"
                class="usa-hint usa-character-count__status"
                data-testid="characterCountMessage"
              >
                2000 characters allowed
              </div>
              <div
                aria-live="polite"
                class="usa-character-count__sr-status usa-sr-only"
                data-testid="characterCountSRMessage"
              />
            </div>
            <div
              class="margin-top-6 margin-bottom-3"
            >
              <button
                class="usa-button usa-button margin-bottom-1"
                data-testid="button"
                type="button"
              >
                Send feedback
              </button>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
